uc-crux-llvm/src/UCCrux/LLVM/FullType/CrucibleType.hs

```diff
@@ -81,55 +99,155 @@ toCrucibleType proxy =
-                          Some fullTypeRepr <-
-                            toFullTypeM (getConst (memTypes Ctx.! idx'))
-                          pure $ Some fullTypeRepr
-                )
+        liftedDecl <-
+          case LLVMTrans.liftDeclare decl of
+            Left err -> throwError (BadLift err)
+            Right d -> pure d
+        Crucible.AnyCFG cfg <-
+          case Map.lookup (L.decName decl) (LLVMTrans.cfgMap trans) of
+            Just (_, anyCfg) -> pure anyCfg
+            Nothing -> throwError (NoCFG (L.decName decl))
+        let crucibleTypes = Crucible.cfgArgTypes cfg
+        let memTypes = fdArgTypes liftedDecl
+        let isVarArgs = fdVarArgs liftedDecl
+        unless
+          ( let numCrucibleTypes = Ctx.sizeInt (Ctx.size crucibleTypes)
+             in length memTypes == numCrucibleTypes
+                  || ( isVarArgs
+                         && length memTypes + 1 == numCrucibleTypes
+                     )
+          )
+          (throwError (BadLiftArgs (L.decName decl)))
+        Some fullTypes <-
+          Ctx.generateSomeM
+            ( Ctx.sizeInt (Ctx.size crucibleTypes)
+                - if isVarArgs
+                  then 1
+                  else 0
+            )
+            ( \idx ->
+                do
+                  Some fullTypeRepr <-
+                    withExceptT FullTypeTranslation $
+                      toFullTypeM (memTypes !! idx)
+                  pure $ Some fullTypeRepr
             )
-            moduleTypes
-        case testCompatibilityAssign proxy fullTypes crucibleTypes of
-          Just Refl -> Right (SomeAssign fullTypes moduleTypes' Refl)
+        retType <-
+          case fdRetType liftedDecl of
+            Just retType ->
+              Just <$> withExceptT FullTypeTranslation (toFullTypeM retType)
+            Nothing -> pure Nothing
+        Some crucibleTypes' <-
+          pure $
+            if not isVarArgs
+              then Some crucibleTypes
+              else case Ctx.viewAssign crucibleTypes of
+                Ctx.AssignEmpty ->
+                  panic
+                    "translateModuleDefines"
+                    ["varargs function with no arguments"]
+                Ctx.AssignExtend rest vaRepr ->
+                  case testEquality vaRepr LLVMTrans.varArgsRepr of
+                    Just Refl -> Some rest
+                    Nothing ->
+                      panic
+                        "translateModuleDefines"
+                        ["varargs function with no varargs repr"]
+        case testCompatibilityAssign (Proxy :: Proxy arch) fullTypes crucibleTypes' of
+          Just Refl ->
+            pure
+              ( L.decName decl,
+                FunctionTypes
+                  { ftArgTypes = MatchingAssign fullTypes crucibleTypes',
+                    ftRetType = retType,
+                    ftIsVarArgs = isVarArgs
+                  }
+              )
           Nothing -> panic "Impossible" []
```

Comment:
Could we say why this is impossible, in case it turns out to not be impossible?

Reply:
Yeah, for sure! Just quickly here: `toCrucibleType . toFullTypeM` should be `id`.